### PR TITLE
TPE: Use `value.get_as_x` methods when available

### PR DESCRIPTION
**Pull request by luxas**
## Description of changes

Broken out from #2162

Using the `get_as_x` methods in TPE minimizes the delta to the concrete evaluator (which uses them), and gives us automatically (later) the right error types for type mismatches (which should not happen if the Residual is validated, but anyways).

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)

I confirm that this PR (choose one, and delete the other options):

- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar language specification.

### CODE REVIEW

File: cedar-policy-core/src/tpe/evaluator.rs
Comment (john-h-kastner-aws) [inline cedar-policy-core/src/tpe/evaluator.rs]: Any reason not to make the same change here?

**Review by john-h-kastner-aws — APPROVED** [review]
Change is good, but I feel like it could probably be more aggressive

File: cedar-policy-core/src/tpe/evaluator.rs
Comment (luxas) [reply cedar-policy-core/src/tpe/evaluator.rs]: I held off at first, as it actually introduced a small change: before, `<residual> && <nonbool>` was left as-is, now `<residual> && <nonbool>` => `<residual> && <typeerror>`. However, this is kinda expected anyways, and should never happen on validated residuals.

Left are only the `Residual::Concrete` matches against `HasAttr` and `GetAttr`, which either match an entity or record, I think it makes the most sense to leave those as-is, as the concrete evaluator also uses similar match statements, and then if none of them match, return a custom `type_error`

**Comment by luxas** [comment]
Answered in https://github.com/cedar-policy/cedar/pull/2323#discussion_r3160152844:

> I held off at first, as it actually introduced a small change: before, <residual> && <nonbool> was left as-is, now <residual> && <nonbool> => <residual> && <typeerror>. However, this is kinda expected anyways, and should never happen on validated residuals.
>
> Left are only the Residual::Concrete matches against HasAttr and GetAttr, which either match an entity or record, I think it makes the most sense to leave those as-is, as the concrete evaluator also uses similar match statements, and then if none of them match, return a custom type_error

However, now I updated the RHS matching for `And` and `Or` to use `get_as_bool` as well, and improved the and and or tests to follow the same order as the match statements, such that we cover all cases exhaustively (some small edge cases were not covered before, I checked the coverage report). Also using insta to assert inner structure of the residual, this could be done for other TPE tests as well in a follow-up.

File: cedar-policy-core/src/tpe/evaluator.rs [hunk 1]
```diff
@@ -97,57 +97,44 @@ impl Evaluator<'_> {
...
+                                // "<left-residual> && <nonbool>" => "<left-residual> && <error>"
+                                // TODO(luxas): Introduce a Residual::PartialError variant that says "the expression definitely errors, but with an unknown error"
+                                Err(_) => mk_residual(ResidualKind::And {
+                                    left: Arc::new(left),
+                                    right: Arc::new(mk_error()),
+                                }),
+                            },
+                            // Cannot simplify "<left-residual> && <right-residual>" or "<left-residual> && <error>"
+                            // The latter expression could become a Residual::PartialError later.
+                            Residual::Partial { .. } | Residual::Error(_) => {
```
Comment (luxas) [inline cedar-policy-core/src/tpe/evaluator.rs:133]: using the explicit notion here instead of `_` for reader clarity.

File: cedar-policy-core/src/tpe/evaluator.rs [hunk 2]
```diff
@@ -830,9 +794,9 @@ mod tests {
                 ..
             },
         );
-        // <non-error-free> && false cannot be simplified, e.g.
-        // principal.foo + 1 == 100 && 41 == 42
-        assert_matches!(
+        // "<non-error-free> && false" cannot be fully simplified, e.g.
+        // "principal.foo + 1 == 100 && 41 == 42" => "principal.foo + 1 == 100 && false"
+        assert_debug_snapshot!(
```
Comment (victornicolet) [inline cedar-policy-core/src/tpe/evaluator.rs]: nit: I'm not a big fan of using debug-print string equality testing on internal representations, because it can add test maintenance burden when the representation changes even a little bit. I'm also saying this based on a very high-level understanding of insta and I don't know if insta makes that easier.

Comment (john-h-kastner-aws) [reply cedar-policy-core/src/tpe/evaluator.rs]: Insta does make this a _lot_ easier. The test update is just running `cargo insta review` which will show you the diff and replace the expected output if you're happy with it.

That said, using the debug repr for expressions does make it difficult to read the tests, so asserting the snapshot on the cedar expression string might be nicer if there isn't anything lost in that conversion (e.g., type information) that we want to assert equality for.

Comment (luxas) [reply cedar-policy-core/src/tpe/evaluator.rs]: Totally, strings are nicer to look at, so asserting those now instead, using insta 😄 
While I was at it, I also converted to use typed string exprs as input, to make it easier for the reader.
I can convert the other tests as well, but preferably in another PR.

**Review by victornicolet — APPROVED** [review]
